### PR TITLE
fix(interp): Coerce function arguments and variable assignments into desired types

### PR DESCRIPTION
**Pull request by vbuchii**
#609
## While running on interpreter:
Screenshot 2021-01-27 at 19 08 40

## On Roku:
Screenshot 2021-01-27 at 19 11 57

### CODE REVIEW

**Comment by vbuchii** [comment]
Hi, @sjbarag 
I've added typecasting but we have different behavior as on Roku. I'm not very familiar with JS and would like to get your feedback on what we should do with that. I skipped a part of the tests and left comments above them. Also, you can find screenshots attached to the PR's description. 
Thanks

**Comment by sjbarag** [comment]
Thanks for getting this started, @vbuchii!  The implementation looks pretty close, so I'm going to pick it up from here 😃 I'll push some commits to this PR, so don't panic if you see some notifications in your email.

File: src/brsTypes/Callable.ts [hunk 1]
```diff
@@ -217,91 +221,126 @@ export class Callable implements Brs.BrsValue {
...
     }
 
+    /**
+     * Attempts to satisfy each signature of this Callable using the provided arguments, coercing arguments into other
+     * types where necessary and possible.
+     * @param args - the arguments to satisfy this Callable with
+     * @returns the signature, implementation, type mismatches, and coerced arguments for each signature in this
+     *          Callable.
+     */
     getAllSignatureMismatches(args: Brs.BrsType[]): SignatureAndMismatches[] {
```
Comment (lkipke) [inline src/brsTypes/Callable.ts:248]: Should this be `SignatureSatisfaction[]`?

File: src/interpreter/index.ts [hunk 2]
```diff
@@ -448,9 +448,33 @@ export class Interpreter implements Expr.Visitor<BrsType>, Stmt.Visitor<BrsType>
...
+                    value.kind === ValueKind.Double)
+            ) {
+                value = new Float(value.getValue());
+            } else if (
+                requiredType === ValueKind.Double &&
+                (value.kind === ValueKind.Float ||
+                    value.kind === ValueKind.Int32 ||
+                    value.kind === ValueKind.Int64)
+            ) {
                 value = new Double(value.getValue());
```
Comment (lkipke) [inline src/interpreter/index.ts:478]: Wondering if it'd be helpful to pull this logic into a separate util function, considering how lengthy/specific it is, and we use it both here and in `trySatisfySignature`.

File: test/brsTypes/Double.test.js [hunk 3]
```diff
@@ -413,4 +413,39 @@ describe("Double", () => {
             expect(six.or(new Double(3.4444))).toEqual(new Int32(7));
         });
     });
+
+    describe("validate typecasting", () => {
+        // If we pass floatValue as argument for Double() thne doubleValue = 'Double { kind: 7, value: Float { kind: 6, value: 10.88889 } }'
+        // But if pass '6.2' then it = 'Double { kind: 7, value: 10.88889 }'
+        // while running in interpreter: Double, value: 10.88889
+        // on Roku:                      Double, value: 10.88888835907
+        xit("passing float", () => {
```
Comment (lkipke) [inline test/brsTypes/Double.test.js:422]: Is there a separate issue surrounding precision that is causing us to skip these tests, or should this PR be fixing these precision issues as well? I don't think i'm fully understanding the above comment

File: test/e2e/TypeChecking.test.js [hunk 4]
```diff
@@ -11,11 +11,11 @@ describe("function argument type checking", () => {
     });
 
     afterEach(() => {
-        stderr.mockClear();
+        jest.resetAllMocks();
```
Comment (lkipke) [inline test/e2e/TypeChecking.test.js:14]: Do we need these changes?

**Review by lkipke — APPROVED** [review]
Overall looks good, just a few comments -- love the extensive tests for this

File: src/interpreter/index.ts (hunk 2 above)
Comment (alimnios72) [reply src/interpreter/index.ts:478]: +1

File: test/brsTypes/Double.test.js [hunk 5]
```diff
@@ -413,4 +413,39 @@ describe("Double", () => {
...
+            let floatValue = new Float(10.88889);
+            let doubleValue = new Double(floatValue);
+
+            expect(doubleValue.equalTo(new Double(10.88888835907))).toBe(BrsBoolean.True);
+        });
+
+        // the same here: 'doubleValue = Double { kind: 7, value: Int32 { kind: 4, value: 20 } }'
+        // while running in interpreter: Double, value: 20
+        // on Roku:                      Double, value: 20
+        xit("passing Integer", () => {
```
Comment (alimnios72) [inline test/brsTypes/Double.test.js:432]: isn't this and below fixed now?

**Review by sjbarag — COMMENTED** [review]
heck my comments from around a month ago never posted 0.o

File: src/brsTypes/Callable.ts [hunk 6]
```diff
@@ -217,91 +222,82 @@ export class Callable implements Brs.BrsValue {
...
+     * @param args - the arguments to satisfy this Callable with
+     * @returns the signature, implementation, type mismatches, and coerced arguments for each signature in this
+     *          Callable.
+     */
+    getAllSignatureMismatches(args: Brs.BrsType[]): SignatureSatisfaction[] {
+        return this.signatures.map((sigAndImpl) => this.trySatisfySignature(sigAndImpl, args));
     }
 
-    private getSignatureMismatches(sig: Signature, args: Brs.BrsType[]): SignatureMismatch[] {
+    private trySatisfySignature(
```
Comment (alimnios72) [inline src/brsTypes/Callable.ts:253]: The name of the function is pretty self explanatory but could you add docs please?

File: test/interpreter/Variables.test.js [hunk 7]
```diff
@@ -80,6 +78,71 @@ describe("interpreter variables", () => {
...
+                ["longinteger", new Int64(2147483647119), new Double(2147483647119)],
+            ])("from %s to integer", (_type, rhs, lhs_result) => {
+                let assign = [
+                    new Stmt.Assignment(tokens, identifier("double#"), new Expr.Literal(rhs)),
+                ];
+                expect(() => interpreter.exec(assign)).not.toThrow();
+                expect(interpreter.environment.get(identifier("double#"))).toEqual(lhs_result);
+            });
+        });
+        describe("into longinteger LHS", () => {
```
[the remark this answers is not in the document]
Comment (alimnios72) [reply test/interpreter/Variables.test.js:129]: @sjbarag please remind @sjbarag to do it

**Review by alimnios72 — APPROVED** [review]
Wow, this wrapped up pretty nicely. Thanks for the hard work here @vbuchii and @sjbarag, these changes will fix lot of the issues I've seen when adding tests

File: test/interpreter/Variables.test.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (sjbarag) [reply test/interpreter/Variables.test.js:129]: > @sjbarag please remind @sjbarag to do it

@sjbarag get your 🐩 together already

**Pull request merged**